halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/validate/v1/ValidatorCollection.java

```diff
@@ -0,0 +1,91 @@
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.halyard.config.validate.v1;
+
+import com.netflix.spinnaker.halyard.config.model.v1.node.Node;
+import com.netflix.spinnaker.halyard.config.model.v1.node.Validator;
+import com.netflix.spinnaker.halyard.config.model.v1.problem.ProblemSetBuilder;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * This collects all validators that have been defined so far, and tries to apply all matching ones to the input node.
+ */
+@Slf4j
+@Component
+public class ValidatorCollection {
+  @Autowired(required = false)
+  private List<Validator> validators = new ArrayList<>();
+
+  /**
+   * Runs every validator defined against the given node.
+   *
+   * @param psBuilder contains the problems encountered during validation so far.
+   * @param node is the node being validated.
+   *
+   * @return # of validators run (for logging purposes).
+   */
+  public int runAllValidators(ProblemSetBuilder psBuilder, Node node) {
+    psBuilder.setNode(node);
+    int validatorRuns = 0;
+    for (Validator validator : validators) {
+      validatorRuns += runMatchingValidators(psBuilder, validator, node, node.getClass());
+    }
+
+    log.info("Total validators run for node " + node.getClass() + " == " + validatorRuns);
+
+    return validatorRuns;
+  }
+
+  /**
+   * Walk up the object hierarchy, running this validator whenever possible. The idea is, perhaps we were passed a
+   * Kubernetes account, and want to run both the standard Kubernetes account validator to see if the kubeconfig is valid,
+   * as well as the super-classes Account validator to see if the account name is valid.
+   *
+   * @param psBuilder contains the list of problems encountered.
+   * @param validator is the validator to be run.
+   * @param node is the subject of validation.
+   * @param c is some super(inclusive) class of node.
+   *
+   * @return # of validators run (for logging purposes).
+   */
+  private int runMatchingValidators(ProblemSetBuilder psBuilder, Validator validator, Node node, Class c) {
+    int result = 0;
+
+    if (c == Node.class) {
+      return result;
+    }
+
+    try {
+      Method m = validator.getClass().getMethod("validate", ProblemSetBuilder.class, c);
+      m.invoke(validator, psBuilder, node);
+      result = 1;
+    } catch (NoSuchMethodException e) {
+      // Do nothing, odds are most validators don't validate every class.
+    } catch (InvocationTargetException | IllegalAccessException e) {
+      log.warn("Failed to invoke validate() on " + validator.getClass() + " for node " + c, e);
+    }
+
+    return runMatchingValidators(psBuilder, validator, node, c.getSuperclass()) + result;
```

Comment:
This feels weird because you are calling validate on each of the classes in the node's hieararchy. Why wouldnt the specialized validate already accomodate this?